src/buffer/out/sgrStack.hpp

```diff
@@ -0,0 +1,238 @@
+/*++
+Copyright (c) Microsoft Corporation
+Licensed under the MIT license.
+
+Module Name:
+- sgrStack.hpp
+
+Abstract:
+- Encapsulates logic for the XTPUSHSGR / XTPOPSGR VT control sequences, which save and
+  restore text attributes on a stack.
+
+--*/
+
+#pragma once
+
+#include "TextAttribute.hpp"
+#include "..\..\terminal\adapter\DispatchTypes.hpp"
+
+namespace Microsoft::Console::VirtualTerminal
+{
+    class SgrStack
+    {
+    public:
+        SgrStack() noexcept :
+            _numSgrPushes{ 0 },
+            _validAttributes{ 0 },
+            _storedSgrAttributes{ 0 }
+        {
+        }
+
+        // Method Description:
+        // - Saves the specified text attributes onto an internal stack.
+        // Arguments:
+        // - currentAttributes - The attributes to save onto the stack.
+        // - options - If none supplied, the full attributes are saved. Else only the
+        //   specified parts of currentAttributes are saved.
+        // Return Value:
+        // - <none>
+        void Push(const TextAttribute& currentAttributes,
+                  const gsl::span<const DispatchTypes::GraphicsOptions> options) noexcept
+        {
+            uint32_t validParts = 0;
+
+            if (options.size() == 0)
+            {
+                // We save all current attributes.
+                validParts = UINT32_MAX;
+            }
+            else
+            {
+                // Each option is encoded as a bit in validParts. Options that aren't
+                // supported are ignored. So if you try to save only unsuppported aspects
+                // of the current text attributes, validParts will be zero, and you'll do
+                // what is effectively an "empty" push (the subsequent pop will not change
+                // the current attributes).
+                for (auto option : options)
+                {
+                    validParts |= _GraphicsOptionToFlag(option);
+                }
+            }
+
+            if (_numSgrPushes < _countof(_storedSgrAttributes))
```

Comment:
is `_countof` portable? Should we use `std::extent<decltype(_storedSgrAttributes)>::value` instead? Man, that's a mouthful!

Reply:
Or we could make it a `std::array` and not have to worry about the mouthful at all 😝